Check for nil before using HostConfig to adjustCpuShares

srust opened the pull request:
Hi,

This fixes #14915 for me.

The remote API in master does not work for a /container/create request that comes in with a missing HostConfig sub-object. Docker will dereference a nil pointer.

The 'adjustCpuShares' function expects a valid HostConfig object.

It appears that the commit for https://github.com/docker/docker/pull/13218 changed the behavior of the decoder, to no longer create a new HostConfig object if none was specified over the API. That change appears to have broken the create call. With this fix applied, the issue no longer occurs.

Thanks!

Signed-off-by: Stephen Rust srust@blockbridge.com

runcom commented:
@srust thanks! Can you add a failing test case for this change also?

runcom commented:
this is a bug where when the client call an older api version w/o hostconfig it results in this nil pointer dereference, failing test case:
```
diff --git a/integration-cli/docker_api_containers_test.go b/integration-cli/docker_api_containers_test.go
index b0e9b0e..48b4fc8 100644
--- a/integration-cli/docker_api_containers_test.go
+++ b/integration-cli/docker_api_containers_test.go
@@ -1687,3 +1687,13 @@ func (s *DockerSuite) TestPostContainersStartWithLinksInHostConfigIdLinked(c *ch
        c.Assert(res.StatusCode, check.Equals, http.StatusNoContent)
        b.Close()
 }
+
+// #14915
+func (s *DockerSuite) TestPostContainersCreateWithoutHostConfigOnOldAPI(c *check.C) {
+       config := struct {
+               Image string
+       }{"busybox"}
+       status, _, err := sockRequest("POST", "/v1.18/containers/create", config)
+       c.Assert(err, check.IsNil)
+       c.Assert(status, check.Equals, http.StatusCreated)
+}
```
ping @calavera I've added this to 1.8 (remove it if I've mistaken)

srust commented:
Ah thanks @runcom for the test case! Was just working through the test suite to add this. I'll move the nil check.

Commit:
Check for nil before using HostConfig to adjustCpuShares

Fix #14915. Add unit test for #14915.
Thanks @runcom for the test case: when the client calls 1.18 api
version w/o hostconfig it results in a nil pointer dereference.

Signed-off-by: Stephen Rust <srust@blockbridge.com>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "14915-empty-hostconfig-on-create" git@github.com:srust/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "14915-empty-hostconfig-on-create" git@github.com:srust/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

runcom commented:
@srust could you also squash your commits down to just one pls?

srust commented:
@runcom I've commited your patch for the unit test (hope that's okay). thanks again. I've pushed both the fix and test case and rebased with master. Let me know anything else you need for this. Thanks!

srust commented:
Squashed to one commit.

runcom commented:
LGTM

cpuguy83 commented on integration-cli/docker_api_containers_test.go:
```diff
@@ -1687,3 +1687,13 @@ func (s *DockerSuite) TestPostContainersStartWithLinksInHostConfigIdLinked(c *ch
 }
+
+// #14915
+func (s *DockerSuite) TestPostContainersCreateWithoutHostConfigOnOldAPI(c *check.C) {
```
`TestContainersApiCreateNoHostConfig118`

srust commented:
Renamed unit test from code review.

cpuguy83 commented:
LGTM

LK4D4 commented:
LGTM

calavera commented:
cherry picked into #15091
